File: engine/profiler.py
import time


class Profiler:
    def __init__(self, game=None, max_samples=180):
        self.game = game
        self.max_samples = max_samples
        self.enabled = True
        self.paused = False
        self.frame_index = 0
        self.system_samples = {}
        self.frame_samples = []
        self.counters = {}
        self._frame_start = None

    def begin_frame(self):
        if not self.enabled or self.paused:
            return

        self._frame_start = time.perf_counter()

    def end_frame(self):
        if not self.enabled or self.paused or self._frame_start is None:
            return

        elapsed = (time.perf_counter() - self._frame_start) * 1000.0
        self.frame_samples.append(elapsed)
        self.frame_samples = self.frame_samples[-self.max_samples:]
        self.frame_index += 1
        self._frame_start = None

    def record_system(self, name, milliseconds):
        if not self.enabled or self.paused:
            return

        samples = self.system_samples.setdefault(name, [])
        samples.append(float(milliseconds))

        if len(samples) > self.max_samples:
            samples.pop(0)

    def set_counter(self, name, value):
        self.counters[name] = value

    def average(self, values):
        if not values:
            return 0.0
        return round(sum(values) / len(values), 3)

    def rows(self):
        rows = [
            ("Frame ms", self.average(self.frame_samples)),
            ("Frame", self.frame_index),
        ]

        for name, samples in sorted(
            self.system_samples.items(),
            key=lambda item: self.average(item[1]),
            reverse=True,
        )[:12]:
            rows.append((name, f"{self.average(samples)} ms"))

        for key, value in sorted(self.counters.items()):
            rows.append((key, value))

        return rows

    def toggle_pause(self):
        self.paused = not self.paused
        return self.paused

    def serialize_snapshot(self):
        return {
            "frame_index": self.frame_index,
            "frame_ms": self.average(self.frame_samples),
            "systems": {
                name: self.average(samples)
                for name, samples in self.system_samples.items()
            },
            "counters": self.counters,
        }

File: src/engine/profiler.rs
use std::collections::BTreeMap;
use std::time::{Duration, Instant};

#[derive(Debug, Clone)]
pub struct Profiler {
    pub frame_start: Option<Instant>,
    pub frame_time: Duration,
    pub systems: BTreeMap<String, f64>,
    pub counters: BTreeMap<String, usize>,
}

impl Default for Profiler {
    fn default() -> Self {
        Self::new()
    }
}

impl Profiler {
    pub fn new() -> Self {
        Self {
            frame_start: None,
            frame_time: Duration::ZERO,
            systems: BTreeMap::new(),
            counters: BTreeMap::new(),
        }
    }

    pub fn begin_frame(&mut self) {
        self.frame_start = Some(Instant::now());
    }

    pub fn end_frame(&mut self) {
        if let Some(start) = self.frame_start.take() {
            self.frame_time = start.elapsed();
        }
    }

    pub fn record_system(&mut self, name: &str, milliseconds: f64) {
        self.systems.insert(name.to_string(), milliseconds);
    }

    pub fn set_counter(&mut self, name: &str, value: usize) {
        self.counters.insert(name.to_string(), value);
    }

    pub fn rows(&self) -> Vec<(String, String)> {
        self.systems
            .iter()
            .map(|(name, value)| (name.clone(), format!("{value:.1} ms")))
            .collect()
    }

    /// Sistema más costoso del último frame (ms).
    pub fn slowest_system(&self) -> Option<(String, f64)> {
        self.systems
            .iter()
            .max_by(|a, b| a.1.partial_cmp(b.1).unwrap_or(std::cmp::Ordering::Equal))
            .map(|(k, v)| (k.clone(), *v))
    }

    pub fn systems_time_total_ms(&self) -> f64 {
        self.systems.values().copied().sum()
    }
}
